be/src/olap/rowset/segment_v2/vertical_segment_writer.cpp

```diff
@@ -655,14 +671,22 @@ Status VerticalSegmentWriter::_append_block_with_variant_subcolumns(RowsInBlock&
             CHECK(entry->data.is_finalized());
             int current_column_id = column_id++;
             TabletColumn tablet_column = generate_column_info(entry);
+            DataTypePtr storage_type =
+                    vectorized::DataTypeFactory::instance().create_data_type(tablet_column);
+            DataTypePtr finalized_type = entry->data.get_least_common_type();
+            vectorized::ColumnPtr current_column =
+                    entry->data.get_finalized_column_ptr()->get_ptr();
+            if (!storage_type->equals(*finalized_type)) {
+                RETURN_IF_ERROR(vectorized::schema_util::cast_column(
```

Comment:
I think you mean casting sub column to user specifified type after normal type processing in variant.
If that's right, I think it's not efficient enough compared to pass type user specifified type to type procession in variant and do the type parsing or casting at first time.